### PR TITLE
Native CLI Distribution

### PR DESCRIPTION
**Pull request by ryanemerson**
Example GitHub release with binaries: https://github.com/ryanemerson/infinispan-quarkus/releases/tag/12.0.0.CR1

@tristantarrant @oraNod The README is very high level, but it's a starting point for `12.0.0.Final`. I think in subsequent versions we need to properly document `kubectl infinispan` in the main docs and simply point to those in the README.

### CODE REVIEW

File: cli/src/main/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1,3 @@
+quarkus.package.output-name=ispn
+# It's not possible to have an empty to suffix, so just use -cli to create the desired ispn-cli
```
Comment (oraNod) [inline cli/src/main/resources/application.properties:2]:
```suggestion
# A suffix is required.  Use the -cli suffix to create the ispn-cli output name.
```

File: cli/src/main/resources/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Infinispan CLI
```
Comment (oraNod) [inline cli/src/main/resources/README.md:1]:
```suggestion
# Infinispan Native CLI
```

File: cli/src/main/resources/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+The Infinispan Command Line Interface (CLI) lets you remotely connect to Infinispan servers to access data and perform administrative functions.
```
Comment (oraNod) [inline cli/src/main/resources/README.md:2]:
```suggestion
The Infinispan Command Line Interface (CLI) connects to Infinispan Server clusters so you can access data remotely and perform administrative operations.
```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    - name: Upload CLI native executable
+      id: upload-cli-native-executable
+      uses: actions/upload-release-asset@v1
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        upload_url:  ${{ needs.release.outputs.upload_url }}
+        asset_path: ./cli/target/infinispan-quarkus-cli-${{ github.event.inputs.releaseVersion }}.zip
+        asset_name: infinispan-cli-${{ github.event.inputs.releaseVersion }}-${{ matrix.asset_name }}
```
Comment (tristantarrant) [inline .github/workflows/release.yml:141]: I think we need to have the `.zip` extension in here too

File: cli/src/main/resources/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+To learn more about the CLI, please visit [infinispan.org](https://infinispan.org/docs/stable/titles/cli/cli.html).
```
Comment (oraNod) [inline cli/src/main/resources/README.md:4]:
```suggestion
You can find complete documentation for the CLI, in our [CLI User Guide](https://infinispan.org/docs/stable/titles/cli/cli.html).
```

File: cli/src/main/resources/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+## Getting Started
+
+#### Linux & Mac
```
Comment (oraNod) [inline cli/src/main/resources/README.md:8]:
```suggestion
#### Linux and macOS
```

File: cli/src/main/resources/README.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+```
+./ispn-cli -c http://<host>:11222
+```
+
+####  Windows
```
Comment (oraNod) [inline cli/src/main/resources/README.md:13]:
```suggestion
####  Microsoft Windows
```

File: cli/src/main/resources/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+Using powershell:
```
Comment (oraNod) [inline cli/src/main/resources/README.md:14]:
```suggestion
Start the CLI with PowerShell and connect to a running Infinispan Server as follows:
```

File: .github/workflows/release.yml (hunk 4 above)
Comment (ryanemerson) [reply .github/workflows/release.yml:141]: Good spot. I forgot to update that after going from binary to zip.

File: cli/src/main/resources/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
+# Infinispan CLI
+The Infinispan Command Line Interface (CLI) lets you remotely connect to Infinispan servers to access data and perform administrative functions.
+
+To learn more about the CLI, please visit [infinispan.org](https://infinispan.org/docs/stable/titles/cli/cli.html).
+
+## Getting Started
+
+#### Linux & Mac
+```
```
Comment (oraNod) [inline cli/src/main/resources/README.md:9]:
```suggestion
Start the CLI and connect to a running Infinispan Server as follows:
```
```

File: cli/src/main/resources/README.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+####  Windows
+Using powershell:
+```
+& ".\ispn-cli.exe" -c http://<host>:11222
+```
+
+## kubectl Integration
+Infinispan CLI can be used as `kubectl` plugin, allowing the CLI to control the Infinispan operator, simplifying many operations.
```
Comment (oraNod) [inline cli/src/main/resources/README.md:20]:
```suggestion
You can use the Infinispan CLI as `kubectl` client plugin, which enables the CLI to control the Infinispan Operator and simplifies many operations.
```

File: cli/src/main/resources/README.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+This includes the ability to install and uninstall the operator, create and delete Infinispan clusters and obtain information about various resources.
```
Comment (oraNod) [inline cli/src/main/resources/README.md:21]:
```suggestion
Integrating with `kubectl` lets you install and remove the Infinispan Operator, create and delete Infinispan clusters, and obtain information about various resources.
```

File: cli/src/main/resources/README.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+Follow the steps below to install the cli as a `kubectl` plugin:
```
Comment (oraNod) [inline cli/src/main/resources/README.md:23]:
```suggestion
To install the CLI as a `kubectl` plugin, do the following:
```

File: cli/src/main/resources/README.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+./ispn-cli -c http://<host>:11222
+```
+
+####  Windows
+Using powershell:
+```
+& ".\ispn-cli.exe" -c http://<host>:11222
+```
+
+## kubectl Integration
```
Comment (oraNod) [inline cli/src/main/resources/README.md:19]:
```suggestion
## Installing the CLI as a kubectl Plugin
```

File: cli/src/main/resources/README.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+Infinispan CLI can be used as `kubectl` plugin, allowing the CLI to control the Infinispan operator, simplifying many operations.
+This includes the ability to install and uninstall the operator, create and delete Infinispan clusters and obtain information about various resources.
+
+Follow the steps below to install the cli as a `kubectl` plugin:
+
+1. Copy the `ispn-cli`, or create a link, to a file with the name "kubectl-infinispan"
```
Comment (oraNod) [inline cli/src/main/resources/README.md:25]:
```suggestion
1. Copy the `ispn-cli`, or create a symbolic link, to a file named "kubectl-infinispan".
```

File: cli/src/main/resources/README.md [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+2. Add `kubectl-infinispan` to your `PATH`
```
Comment (oraNod) [inline cli/src/main/resources/README.md:26]:
```suggestion
2. Add `kubectl-infinispan` to your `PATH`.
```

File: cli/src/main/resources/README.md [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+3. Verify the plugin has been detected by executing `kubectl plugin list`
```
Comment (oraNod) [inline cli/src/main/resources/README.md:27]:
```suggestion
3. Run `kubectl plugin list` to verify the CLI is added as a plugin. 
```

File: cli/src/main/resources/README.md [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+4. Start using the plugin `kubectl infinispan --help`
```
Comment (oraNod) [inline cli/src/main/resources/README.md:28]:
```suggestion
4. Run `kubectl infinispan --help` to start using the CLI plugin.
```

File: cli/src/main/resources/README.md [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+For more detailed instructions, please see the [kubectl docs](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#installing-kubectl-plugins).
+
+#### Operator Installation
```
Comment (oraNod) [inline cli/src/main/resources/README.md:32]:
```suggestion
#### Installing Infinispan Operator
```

File: cli/src/main/resources/README.md [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+```
+kubectl infinispan install
+```
+
+#### Cluster creation
```
Comment (oraNod) [inline cli/src/main/resources/README.md:37]:
```suggestion
#### Creating Infinispan Clusters
```

File: cli/src/main/resources/README.md [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+```
+kubectl infinispan create --replicas=2 myinfinispan
+kubectl infinispan ls
+```
+
+#### Cluster Removal
```
Comment (oraNod) [inline cli/src/main/resources/README.md:43]:
```suggestion
#### Removing Clusters
```

File: cli/src/main/resources/README.md [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+```
+kubectl infinispan rm myinfinispan
+kubectl infinispan ls
+```
+
+#### Operator Removal
```
Comment (oraNod) [inline cli/src/main/resources/README.md:49]:
```suggestion
#### Removing the Operator
```

File: cli/src/main/resources/README.md [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+#### Operator Installation
+```
+kubectl infinispan install
+```
+
+#### Cluster creation
+```
+kubectl infinispan create --replicas=2 myinfinispan
+kubectl infinispan ls
+```
```
Comment (oraNod) [inline cli/src/main/resources/README.md:41]: @ryanemerson Does this create cache service clusters by default? Should we maybe show how to create DG clusters as well?

File: .github/workflows/release.yml [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+
+    - name: Checkout Release Tag
+      uses: actions/checkout@v2
+      with:
+        ref: ${{ github.event.inputs.releaseVersion }}
+
+    - if: ${{ matrix.os == 'windows-latest' }}
+      name: Build native executable
+      run: |
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x86_amd64
```
Comment (wburns) [inline .github/workflows/release.yml:125]: Is this always going to be this path? Is there not a variable we can use or something?

**Review by wburns — COMMENTED** [review]
Just a minor thing, most of this is black magic stuff for me, so 👍

File: .github/workflows/release.yml (hunk 23 above)
Comment (ryanemerson) [reply .github/workflows/release.yml:125]: It should always be the case when VS 2019 is used ... not sure after that. I don't know of a better way of doing this, or if a variable exists, tbh the windows build was nothing but hassle.

File: cli/src/main/resources/README.md (hunk 22 above)
Comment (ryanemerson) [reply cli/src/main/resources/README.md:41]: Good point, but AFAIK that's not actually possible with the CLI atm.

**Comment by ryanemerson** [comment]
Self-merged as required for release

**Pull request merged**